### PR TITLE
Port the project to C++11 - Part 1

### PR DESCRIPTION
**Pull request by lzaoral**
This PR ports `src/abstract-*` to C++ 11.

### CODE REVIEW

File: src/abstract-parser.hh [hunk 1]
```diff
@@ -55,10 +55,10 @@ class AbstractParser {
         }
 
     protected:
-        AbstractParser() { }
+        AbstractParser() = default;
 
     private:
-        const TScanProps emptyProps_;
+        const TScanProps emptyProps_{};
```
Comment (kdudka) [inline src/abstract-parser.hh:61]: Just curious, why do we need the extra brackets?

Comment (lzaoral) [reply src/abstract-parser.hh:61]: Defaulted constructors do not initialize non-`static` and `const` members. From https://en.cppreference.com/w/cpp/language/default_constructor:

>The implicitly-declared or defaulted (since C++11) default constructor for class T is defined as deleted if any of the following is true:
>
>    T has a non-[const-default-constructible](https://en.cppreference.com/w/cpp/language/default_initialization#Default_initialization_of_a_const_object) const member without a default member initializer (since C++11).

TL;DR non-`static` but `const` members of types with non-user provided default constructors won't be initialized by the defaulted default constructor so you have to default-initialize them explicitly by default member initializer.

Yeah, C++ is a nice language. 🙃

File: src/abstract-filter.cc [hunk 2]
```diff
@@ -186,7 +186,6 @@ void CtxEmbedder::handleDef(const Defect &defOrig)
     dropCtxLines(&def.events);
     appendCtxLines(&def.events, fstr, evt.line, ctxLines_ - 1);
 
-    // close the file stream and forward the result
-    fstr.close();
```
Comment (kdudka) [inline src/abstract-filter.cc:190]: I do not think the call to `fstr.close()` is redundant.  If you remove it, the file will be unnecessarily kept open until `agent_->handleDef(def)` finishes.  There can be a big chain of decorators behind the call.

File: src/abstract-tree.hh [hunk 3]
```diff
@@ -35,12 +35,8 @@ class AbstractTreeDecoder {
...
         /// read scan properties if available
         virtual void readScanProps(
-                TScanProps             *pDst,
-                const pt::ptree        *root)
-        {
-            (void) pDst;
-            (void) root;
-        }
+                TScanProps      * /* pDst */,
+                const pt::ptree * /* root */) { }
```
Comment (kdudka) [inline src/abstract-tree.hh:39]: I would remove this change.  The previous syntax feels more readable to me and did not cause any problems, as far as I know.

File: src/abstract-filter.hh [hunk 4]
```diff
@@ -57,7 +57,7 @@ class GenericAbstractFilter: public AbstractWriter {
 /// decorator
 class EventPrunner: public GenericAbstractFilter {
     private:
-        int thr_;
+        unsigned thr_;
```
Comment (kdudka) [inline src/abstract-filter.hh:60]: This change triggers a new warning:
```
src/abstract-filter.cc: In member function ‘virtual void EventPrunner::handleDef(const Defect&)’:
src/abstract-filter.cc:87:32: warning: comparison of integer expressions of different signedness: ‘const int’ and ‘unsigned int’ [-Wsign-compare]
   87 |         if (evt.verbosityLevel <= thr_)
      |             ~~~~~~~~~~~~~~~~~~~^~~~~~~
```

**Review by kdudka — CHANGES_REQUESTED** [review]
Thank you for working on this!  Overall it looks good.  A few minor changes suggested inline.

File: src/abstract-filter.hh (hunk 4 above)
Comment (kdudka) [reply src/abstract-filter.hh:60]: I think the proper fix would be this:
```diff
--- a/src/abstract-filter.hh
+++ b/src/abstract-filter.hh
@@ -61,7 +61,7 @@ class EventPrunner: public GenericAbstractFilter {
         int thr_;

     public:
-        EventPrunner(AbstractWriter *agent, unsigned thr):
+        EventPrunner(AbstractWriter *agent, int thr):
             GenericAbstractFilter(agent),
             thr_(thr)
         {
```

File: src/abstract-filter.cc (hunk 2 above)
Comment (lzaoral) [reply src/abstract-filter.cc:190]: I did not take into account this possibility. I'll revert this change.

File: src/abstract-tree.hh (hunk 3 above)
Comment (lzaoral) [reply src/abstract-tree.hh:39]: Yes, it's only a cosmetic change but I think it makes the code more concise.

File: src/abstract-filter.cc (hunk 2 above)
Comment (lzaoral) [reply src/abstract-filter.cc]: Change dropped.

File: src/abstract-tree.hh (hunk 3 above)
Comment (lzaoral) [reply src/abstract-tree.hh]: Commit dropped,

File: src/abstract-filter.hh (hunk 4 above)
Comment (lzaoral) [reply src/abstract-filter.hh]: Addressed in 1fd0a96.

File: src/csparser.hh [hunk 5]
```diff
@@ -30,6 +30,10 @@ class CovParser: public AbstractParser {
         bool getNext(Defect *);
         bool hasError() const;
 
+        virtual EFileFormat inputFormat() const {
```
Comment (kdudka) [inline src/csparser.hh:33]: Should we spell this as `EFileFormat inputFormat() const override {` to be consistent with commit ecd971c4b951ed5b54db5ad5648b6f9f7d9c53ee?

File: src/gcc-parser.hh [hunk 6]
```diff
@@ -30,6 +30,10 @@ class GccParser: public AbstractParser {
         bool getNext(Defect *);
         bool hasError() const;
 
+        virtual EFileFormat inputFormat() const {
```
Comment (kdudka) [inline src/gcc-parser.hh:33]: ditto

File: src/json-parser.hh [hunk 7]
```diff
@@ -31,6 +31,10 @@ class JsonParser: public AbstractParser {
         virtual bool hasError() const;
         virtual const TScanProps& getScanProps() const;
 
+        virtual EFileFormat inputFormat() const {
```
Comment (kdudka) [inline src/json-parser.hh:34]: ditto

**Review by kdudka — APPROVED** [review]

File: src/csparser.hh (hunk 5 above)
Comment (lzaoral) [reply src/csparser.hh:33]: Definitely! I want to make this change as well but in a separate PR as this one is too big already (and focuses mainly on `src/abstrac-*` related changes). Is that ok with you?

Comment (kdudka) [reply src/csparser.hh:33]: This is new code you are introducing in this pull request.  I would prefer use the desired form since the beginning, rather than adding new code that we already plan to rework in a future pull request.  If it depends on other changes, the commit in question can be moved to the next pull request, to avoid unnecessary changes back and forth.

Comment (lzaoral) [reply src/csparser.hh:33]: Ok, let's drop this for now.

**Comment by lzaoral** [comment]
Thanks for the reviews, @kdudka!

**Comment by kdudka** [comment]
No problem.  Thank you for modernizing the code!
